std/file.d

```diff
@@ -465,30 +467,98 @@ version (linux) @safe unittest
 
-Throws: $(D FileException) on file error, $(D UTFException) on UTF
-decoding error.
- */
+    Returns: Array of characters read.
 
-S readText(S = string, R)(R name)
-if (isSomeString!S &&
-    (isInputRange!R && !isInfinite!R && isSomeChar!(ElementEncodingType!R) || isSomeString!R) &&
-    !isConvertibleToString!R)
+    Throws: $(LREF FileException) if there is an error reading the file,
+            $(REF BOMException, std, encoding) if there is a BOM at the front
+            of the input, and it does not match the requested encoding, and
+            $(REF UTFException, std, utf) if the text contains invalid Unicode.
++/
+S readText(S = string, R)(auto ref R name)
+if (isSomeString!S && (isInputRange!R && !isInfinite!R && isSomeChar!(ElementType!R) || is(StringTypeOf!R)))
 {
-    import std.utf : validate;
+    import std.algorithm.searching : startsWith;
+    import std.encoding : BOM, BOMException;
+    import std.exception : enforce;
+    import std.format : format;
+    import std.utf : UTFException, validate;
+
+    static if(is(StringTypeOf!R))
+        StringTypeOf!R filename = name;
+    else
+        auto filename = name;
+
+    immutable bomSeq = readBOM(filename);
+    immutable bom = bomSeq.schema;
+
+    static if(is(Unqual!(ElementEncodingType!S) == char))
+    {
+        with(BOM) switch(bom)
+        {
+            case none:
+            case utf8: break;
+            default: throw new BOMException(format!"UTF-8 requested, but BOM is %s"(bom), bomSeq);
+        }
+    }
+    else static if(is(Unqual!(ElementEncodingType!S) == wchar))
+    {
+        with(BOM) switch(bom)
+        {
+            case none:
+            version(BigEndian)
+            {
+                case utf16be: break;
+                default: throw new BOMException(format!"UTF-16 BE requested, but BOM is %s"(bom), bomSeq);
+            }
+            else
+            {
+                case utf16le: break;
+                default: throw new BOMException(format!"UTF-16 LE requested, but BOM is %s"(bom), bomSeq);
+            }
+        }
+    }
+    else
+    {
+        with(BOM) switch(bom)
+        {
+            case none:
+            version(BigEndian)
+            {
+                case utf32be: break;
+                default: throw new BOMException(format!"UTF-32 BE requested, but BOM is %s"(bom), bomSeq);
+            }
+            else
+            {
+                case utf32le: break;
+                default: throw new BOMException(format!"UTF-32 LE requested, but BOM is %s"(bom), bomSeq);
+            }
+        }
+    }
+
+    auto data = read(filename);
+    if (data.length % ElementEncodingType!S.sizeof != 0)
+        throw new UTFException(format!"The content of %s is not UTF-%s" (filename, ElementEncodingType!S.sizeof * 8));
+
     static auto trustedCast(void[] buf) @trusted { return cast(S) buf; }
-    auto result = trustedCast(read(name));
+    auto result = trustedCast(data[bomSeq.sequence.length .. $]);
```

Comment:
Stripping the BOM may be a breaking change in some situations, e.g. if the code already expects the BOM and will do `s = fn.readText[3..$];`, so I'm not sure about this change.

Reply:
It's my understanding that it's never actually guaranteed that there will be a BOM (though I think that it's usually there for UTF-16). So, as long as existing code is checking that there is a BOM, and removing it if present, then it will be fine. It's only code that simply assumes that a BOM is present that would be in trouble, and that code would be wrong anyway. And it's far more user friendly to strip it off. Otherwise, technically, everyone needs to always be calling `getBOM` to see if there's a BOM and then strip it off, otherwise they have junk at the beginning of the text some percentage of the time. The main reason that that doesn't happen more is that UTF-8 doesn't usually have a BOM.

So, given that anyone handling the BOM now would need to be checking for it and then stripping it if present, the fact that it's removed shouldn't cause problems. It just means that the existing checks would become unnecessary.

Reply:
> It's my understanding that it's never actually guaranteed that there will be a BOM (though I think that it's usually there for UTF-16). 

Err, that completely depends on the use case.

> It's only code that simply assumes that a BOM is present that would be in trouble, and that code would be wrong anyway. 

No, that lies on the assumption that `readText` will only ever be used to read arbitrary files with no foreknowledge of what they may contain. In contrast, consider a program that creates and then reads its own text files: it will put a BOM at the beginning when writing the file, so it will want to remove it when reading it.

> And it's far more user friendly to strip it off.

Well, I don't disagree. I was actually a little surprised the initial `readText` implementation didn't do this, years ago.

On the other hand, that would break the following contract, which might be surprising:

- `fileName.readText.toFile(fileName)` should not change the contents of `fileName`.

Hypothetically, there might be a program that needs to process a large number of text files, do some processing that would affect 1% of them, and then commit them to git or something; a change in `readText` as done here would mean that 100% files will change instead of that 1%.

> Otherwise, technically, everyone needs to always be calling getBOM to see if there's a BOM and then strip it off, otherwise they have junk at the beginning of the text some percentage of the time.

I think a `stripBOM` helper would be much better than `getBOM`, i.e.:

```d
string fileContent = fileName.readText.stripBOM;
```

> The main reason that that doesn't happen more is that UTF-8 doesn't usually have a BOM.

Yeah, I noticed the `dos2unix` utility will remove the BOM on the files it processes by default. HTML and PHP also both don't recommend using a BOM.

Considering the BOM is not used a lot, maybe it's better to err on the side of caution than a breaking change without a significant benefit.

Reply:
Hmm. The main advantage to `getBOM` would be that you can determine the encoding ahead of time. Then if it's not UTF-8, UTF-16, or UTF-32, you can do whatever you need to do with `read` to process it, and if it's UTF-8, UTF-16, or UTF-32, you can call the `readText` with the correct string type. Having `stripBOM` would allow us to avoid having `readText` change the file contents for those cases where it matters, though that's less user-friendly for the common case. Stripping it mostly just matters for UTF-16 though, since UTF-8 almost never has a BOM, and UTF-32 isn't used much. So, I guess that it would be better to have `stripBOM` than to have `readText` strip it off, though that is a bit annoying for any use case I've had. I think that `readBOM` still has value though in and of itself.

Reply:
> I think that readBOM still has value though in and of itself.

Not saying it's not useful, but I think we should avoid encouraging patterns in the standard library that are susceptible to filesystem race conditions (i.e. TOCTOU).

Reply:
I would not expect there to be much risk of a file suddenly being rewritten with a different encoding, and how on earth else are you going to deal with this? Call `readText` and have it fail with the requested encoding after reading in the entire file and then read in the entire file again with its proper encoding? You have the same race condition in that case, but you ended up reading the entire file twice.

Reply:
I think that if you want your program to read UTF-8, UTF-16 and UTF-32 and then process each of those as `string`, `wstring` and `dstring` respectively, you should read the file as an `ubyte` array first, look at the BOM, and then cast the `ubyte[]` to the corresponding string type.

I don't know if there's much demand for it, but the above sounds like something that ought to be handled by the standard library, not user code...

Reply:
Unless we're going to just convert the entire thing to the desired string type regardless of the original encoding, I don't see how a library function could wrap that, because the return type would differ - though unless you've templatized all of your string processing, you're going to have to convert it to a specific string type anyway, or your code won't be able to handle it. I don't think that there's truly a clean way to handle it though. :|

I have started writing a function which is supposed to return a range rather than a string, and the range will take whatever the original encoding was (at least, out of UTF-8, UTF-16, and UTF-32) and convert it to the requested encoding, but it's a bit tricky, especially if the file doesn't have a BOM. It's too bad that we can't just assume UTF-8 for everything...

Reply:
> Unless we're going to just convert the entire thing to the desired string type regardless of the original encoding, I don't see how a library function could wrap that, because the return type would differ - though unless you've templatized all of your string processing, you're going to have to convert it to a specific string type anyway, or your code won't be able to handle it. I don't think that there's truly a clean way to handle it though. :|

Err, yes, I meant eagerly convert it to the return type, which would probably still be a negligible cost for most use cases. But we can keep the type too, by allowing the user to pass a lambda, i.e.: `fileName.readText!(s => s.retro.toFile("reversed.txt"));`